File: shared/src/test/scala/stasis/test/specs/unit/shared/interop/InteropResources.scala
package stasis.test.specs.unit.shared.interop

import scala.io.Source
import scala.util.Using

import play.api.libs.json.JsValue
import play.api.libs.json.Json

object InteropResources {
  def load(domain: String, resource: String): JsValue = {
    val path = s"/interop/$domain/$resource.json"

    Using.resource(
      Option(getClass.getResourceAsStream(path))
        .getOrElse(throw new IllegalStateException(s"Interop resource not found: [$path]"))
    ) { stream =>
      Json.parse(Source.fromInputStream(stream).mkString)
    }
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/interop/InteropResources.kt
package stasis.test.client_android.lib.interop

import com.squareup.moshi.Moshi

object InteropResources {
    fun load(domain: String, resource: String): String {
        val path = "/interop/$domain/$resource.json"
        val stream = checkNotNull(InteropResources::class.java.getResourceAsStream(path)) {
            "Interop resource not found: [$path]"
        }

        return stream.use { it.readBytes().toString(Charsets.UTF_8) }
    }

    fun tree(json: String): Any? =
        normalize(Moshi.Builder().build().adapter(Any::class.java).fromJson(json))

    private fun normalize(value: Any?): Any? =
        when (value) {
            is Map<*, *> -> value
                .filterKeys { it !in ignoredKeys }
                .filterValues { it != null }
                .mapValues { normalize(it.value) }
            is List<*> -> value.map { normalize(it) }
            else -> value
        }

    private val ignoredKeys: Set<Any?> = setOf(
        "next_invocation",
        "use_query_string",
        "context",
        "additional_config"
    )
}
